app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,313 @@
+            spatialCache.replaceAllInBBox(addedOrUpdatedElements.filterIsInstance<Node>(), bbox)
+        }
+
+        synchronized(this) {
+            // first delete
+            deletedKeys.forEach {
+                wayRelationCache.remove(it)
+                wayRelationGeometryCache.remove(it)
+                if (it.type == ElementType.NODE) wayIdsByNodeIdCache.remove(it.id)
+                relationIdsByElementKeyCache.remove(it)
+            }
+            // then add
+            val nodeIds = (addedOrUpdatedElements.filterIsInstance<Node>().map { it.id } + spatialCache.getKeys()).toSet()
+            val ways = addedOrUpdatedElements.filterIsInstance<Way>()
+            val wayIds = ways.map {it.id}
+            val relations = addedOrUpdatedElements.filterIsInstance<Relation>()
+            val relationIds = relations.map {it.id}
+            (ways + relations).forEach { wayRelationCache[ElementKey(it.type, it.id)] = it }
+            addedOrUpdatedGeometries.forEach {
+                if (it.elementType != ElementType.NODE)
+                    wayRelationGeometryCache[ElementKey(it.elementType, it.elementId)] = it.geometry
+            }
+
+            ways.forEach { way ->
+                wayRelationCache[ElementKey(ElementType.WAY, way.id)]?.let { oldWay ->
+                    // remove old way from wayIdsByNodeIdCache
+                    (oldWay as Way).nodeIds.forEach {
+                        wayIdsByNodeIdCache[it]?.remove(way.id)
+                    }
+                }
+                way.nodeIds.forEach {
+                    if (it in nodeIds)
+                        wayIdsByNodeIdCache.getOrPut(it) { ArrayList(2) }.add(way.id)
+                }
+            }
+            relations.forEach { relation ->
+                wayRelationCache[ElementKey(ElementType.RELATION, relation.id)]?.let { oldRelation ->
+                    // remove old relation from relationIdsByElementKeyCache
+                    (oldRelation as Relation).members.forEach {
+                        relationIdsByElementKeyCache[ElementKey(it.type, it.ref)]?.remove(relation.id)
+                    }
+                }
+                relation.members.forEach {
+                    if ((it.ref in nodeIds && it.type == ElementType.NODE)
+                        || (it.ref in wayIds && it.type == ElementType.WAY)
+                        || (it.ref in relationIds && it.type == ElementType.RELATION))
+                        relationIdsByElementKeyCache.getOrPut(ElementKey(it.type, it.ref)) { ArrayList(2) }.add(relation.id)
+                }
+            }
+
+        }
+
+    }
+
+    fun getElement(type: ElementType, id: Long, fetch: (ElementType, Long) -> Element?): Element? {
+        val element = if (type == ElementType.NODE) spatialCache.get(id)
+        else synchronized(this) { wayRelationCache.getOrPutIfNotNull(ElementKey(type, id)) { fetch(type, id) } }
```

Comment:
The format had me confused for a bit

```suggestion
        val element = 
            if (type == ElementType.NODE) spatialCache.get(id)
            else synchronized(this) { wayRelationCache.getOrPutIfNotNull(ElementKey(type, id)) { fetch(type, id) } }
```